### ISSUE TITLE
`multiplot` layout question

### ISSUE THREAD

**Issue by c-alpha** [debug-ito/Gnuplot-Builder#4]
I am arranging plots in a multiplot in two columns:
```perl
  $builder->multiplot(
    'layout ' . $plot_rows . ',2 columnsfirst',
    sub {
      my $inner_builder = Gnuplot::Builder::Script->new;
      ...
    }
  );
```
The two columns are not always filled up like in this example: [V4_2.pdf](https://github.com/debug-ito/Gnuplot-Builder/files/12731975/V4_2.pdf)

This means that I need to use a `set multiplot next` one or more times at some point.

**What's the best or foreseen way of doing this?**

I have made a few experiments, and ended up with:
```perl
$inner_builder->run( "set multiplot next" );
```
in the `sub` passed to `$builder->multiplot`.

**Comment by c-alpha** [comment]
Toshio Ito ***@***.***> writes:

> [...]
> In order to always write a single command string to the gnuplot
> process, use the writer object passed to the code-ref. For example,
>
> ```perl
> $builder->multiplot(
>     "layout 3,2 columnsfirst",
>     sub {
>         my $writer = shift;
>         my $inner_builder = Gnuplot::Builder::Script->new;
>         ... (1st plot) ...
>         $writer->("set multiplot next\n");
>         ... (2nd plot) ...
>         ... (3rd plot) ...
>         $writer->("set multiplot next\n");
>         ... (4th plot) ...
>     }
> )
> ```
> [...]

Thanks for the hint! That's kind of what I guessed after reading the
documentation of `multiplot($option, $code)`, but I got distracted by
the very last sentence of the text there:

> So, if you stick to using Gnuplot::Builder::Script in the `$code`, you
> don't need to use `$writer` explicitly.

This somehow led me to think that not using `$writer` explicitly was
somehow preferable.

OTOH, mixing the use of the inner builder object and the `$writer`
passed into the anonymous `sub` as in your suggested solution, can be
perceived as mixing different layers of abstraction, and thus doesn't
seem to feel quite right. Also, some people (with my humble self being a
prime example) will likely forget adding the newline `\n` at the end. ;-)

What do you think about adding a new convenience function (in the spirit
of `print` vs. `say`):
```perl
$inner_builder->say('set multiplot next');
```
This would seem like a quite intuitive way to me.

Just my two cents anyway, and looking forward to your thoughts.

**Comment by debug-ito** [comment]
> This somehow led me to think that not using `$writer` explicitly was
> somehow preferable.

Sorry about that. I just couldn't imagin your use case. I didn't even know `set multiplot next` command. I'll update that part of the doc.

> mixing different layers of abstraction, and thus doesn't seem to feel quite right. 

I understand, but in the design of Gnuplot::Builder, it's perfectly ok to mix them. That's because it's clearly documented how the Script object uses `$writer` to input data to the gnuplot process.

In my opinion, `Gunplot::Builder::Script` object is an abstraction of a gnuplot script, that is, a series of gnuplot commands to make (basically a single) plot.
The Script object is NOT an abstraction of a channel to the gnuplot process. The channel of the gnuplot process is expressed as the `$writer` object.

The `set multiplot next` command doesn't belong to any plot script, I think. It's like a separator between plots. That's why I think it's quite natural to use `$writer` directly, instead of using a Script object.

That said...

> some people (with my humble self being aprime example) will likely forget adding the newline `\n` at the end.

I totally agree with you. `$writer` is still too low level interface.

So, I think I just come up with another way to achieve the goal.

```perl
$builder->multiplot(
    "layout 3,2 columnsfirst",
    sub {
        Gnuplot::Builder::Script->new->run(
            sub {
               my $inner_builder = Gnuplot::Builder::Script->new;
               ... (1st plot) ...
            },
            "set multiplot next",
            sub {
               ... (2nd plot) ...
            },
            sub {
               ... (3rd plot) ...
            },
            "set multiplot next",
        );
    }
);
```

Looks like command strings passed to `run` method will be suffixed with `\n` properly.

If the above example is too complicated, probably I can extend the API spec of `multiplot` and `multiplot_with`, so that we could write like

```perl
$builder->multiplot(
    "layout 3,2 columnsfirst",
    sub {
       my $inner_builder = Gnuplot::Builder::Script->new;
       ... (1st plot) ...
    },
    "set multiplot next",
    sub {
       ... (2nd plot) ...
    },
    sub {
       ... (3rd plot) ...
    },
    "set multiplot next",
);
```

**Comment by c-alpha** [comment]
Toshio Ito ***@***.***> writes:

> [...]
> In my opinion, `Gunplot::Builder::Script` object is an abstraction of
> a gnuplot script, that is, a series of gnuplot commands to make
> (basically a single) plot. The Script object is NOT an abstraction of
> a channel to the gnuplot process. The channel of the gnuplot process
> is expressed as the `$writer` object.

Ah, that explains; thanks for the clarification. It might seem
worthwhile adding this paragraph of yours to the documentation, too?

> The `set multiplot next` command doesn't belong to any plot script, I
> think. It's like a separator between plots. 

It's a separator between plots in a `multiplot`.

It seems to me that Gnuplot is showing a bit of history here; likely,
the `multiplot` mechanism was added much later. Hence it can appear to
"break" the mental model of what a plot is.

> [...]
>> some people (with my humble self being aprime example) will likely
>> forget adding the newline `\n` at the end.
>
> I totally agree with you. `$writer` is still too low level interface.
>
> So, I think I just come up with another way to achieve the goal.
>
> ```perl
> $builder->multiplot(
>     "layout 3,2 columnsfirst",
>     sub {
>         Gnuplot::Builder::Script->new->run(
>             sub {
>                my $inner_builder = Gnuplot::Builder::Script->new;
>                ... (1st plot) ...
>             },
>             "set multiplot next",
>             sub {
>                ... (2nd plot) ...
>             },
>             sub {
>                ... (3rd plot) ...
>             },
>             "set multiplot next",
>         );
>     }
> );
> ```
> [...]

Unfortunately this would not work in my case. I have two columns of
plots, and either of the columns may not be fully occupied by plots,
depending on the data I am processing. That is when I need "multiplot
next" to skip over the empty layout slots of the multiplot. The number
of plots in each column is dynamic, and can be determined at run-time
only. Thus, the number of times I need to "multiplot next" can also be
computed at run-time only.

Therefore, my current code looks something like this:
```perl
  my $num_in    = datafile_num_channels($infile);
  my $num_out   = datafile_num_channels($outfile);
  my $plot_rows = ( $num_in > $num_out ) ? $num_in : $num_out;

  $builder->multiplot(
    'layout ' . $plot_rows . ',2 columnsfirst',
    sub ($writer) {
      my $inner_builder = Gnuplot::Builder::Script->new;

      # left column of plots (inputs)
      for my $i ( 0 .. ( $num_in - 1 ) ) {
        $inner_builder->set( title => '"Device Input ' . $i . '"' , ... );
        $dataset_in_A->set( index => $i, using => '1:($2)', ... );
        $dataset_in_B->set( index => $i, using => '1:($2-$3)',... );
        $inner_builder->plot( $dataset_in_A, $dataset_in_B );
      }
      for ( 1 .. ( $plot_rows - $num_in ) ) {
        $writer->( "set multiplot next\n" );
      }

      # right column of plots (outputs)
      for my $i ( 0 .. ( $num_out - 1 ) ) {
        $inner_builder->set( title => '"Device Output ' . $i . '"', ... );
        $dataset_out_A->set( index => $i, using => '1:($2)', ... );
        $dataset_out_B->set( index => $i, using => '1:($2-$3)',... );
        $inner_builder->plot( $dataset_out_A, $dataset_out_B );
      }
      for ( 1 .. ( $plot_rows - $num_out ) ) {
        $writer->( "set multiplot next\n" );
      }
    }
  );

```

This would be hard to achieve with your new API, since we can not have a
bare `"set multiplot next"` string as the body of a `for` loop.

With your clarification of `Gunplot::Builder::Script` (a Gnuplot script)
vs. `$writer` (a channel to a Gnuplot process) added to the
documentation, my current code is INHO both, more easily understandable,
and more flexible than your proposed new API. So I think the
clarification of the documentation is all that would be needed in my view.

**Comment by debug-ito** [comment]
> The number of plots in each column is dynamic,

I see. That must be a complicated situation.

> So I think the clarification of the documentation is all that would be needed in my view.

Thanks for understanding. I'll update the document. Maybe I'll ask you for review.

**Comment by debug-ito** [comment]
Fixed by #5
